File: app/src/main/java/com/travlog/android/apps/ui/activities/MyPageActivity.java
package com.travlog.android.apps.ui.activities;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.AppCompatImageView;
import android.view.View;
import android.widget.TextView;

import com.jakewharton.rxbinding2.view.RxView;
import com.travlog.android.apps.R;
import com.travlog.android.apps.libs.BaseActivity;
import com.travlog.android.apps.libs.glide.GlideApp;
import com.travlog.android.apps.libs.qualifiers.RequiresActivityViewModel;
import com.travlog.android.apps.viewmodels.MyPageViewModel;

import butterknife.BindView;
import io.reactivex.android.schedulers.AndroidSchedulers;

import static com.bumptech.glide.request.RequestOptions.circleCropTransform;

@RequiresActivityViewModel(MyPageViewModel.class)
public class MyPageActivity extends BaseActivity<MyPageViewModel> {

    @BindView(R.id.settings)
    View settingsButton;
    @BindView(R.id.profile_picture)
    AppCompatImageView profilePictureImage;
    @BindView(R.id.username)
    TextView usernameText;

    @SuppressLint("CheckResult")
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.a_my_page);

        RxView.clicks(settingsButton)
                .compose(bindToLifecycle())
                .subscribe(__ -> this.startSettingsActivity());

        RxView.clicks(usernameText)
                .compose(bindToLifecycle())
                .subscribe(__ -> this.startSetUserNameActivity());

        viewModel.outputs.setProfilePicture()
                .compose(bindToLifecycle())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::setProfilePicture);

        viewModel.outputs.setUsernameText()
                .compose(bindToLifecycle())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::setUsernameText);
    }

    private void setProfilePicture(final @NonNull String profilePicture) {
        GlideApp.with(this)
                .load(profilePicture)
                .apply(circleCropTransform())
                .into(profilePictureImage);
    }

    private void setUsernameText(final @NonNull String username) {
        this.usernameText.setText(username);
    }

    private void startSettingsActivity() {
        final Intent intent = new Intent(this, SettingsActivity.class);
        startActivityWithTransition(intent, R.anim.slide_in_right, R.anim.fade_out_slide_out_left);
    }

    private void startSetUserNameActivity() {
        final Intent intent = new Intent(this, SetUsernameActivity.class);
        startActivityWithTransition(intent, R.anim.slide_in_right, R.anim.fade_out_slide_out_left);
    }
}

File: app/src/main/java/com/travlog/android/apps/ui/activities/MyPageActivity.kt
/*
 * Copyright 2019 Travlog. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.travlog.android.apps.ui.activities

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import com.bumptech.glide.request.RequestOptions.circleCropTransform
import com.jakewharton.rxbinding3.view.clicks
import com.travlog.android.apps.R
import com.travlog.android.apps.ViewModelFactory
import com.travlog.android.apps.getAppInjector
import com.travlog.android.apps.getViewModel
import com.travlog.android.apps.libs.BaseActivity
import com.travlog.android.apps.libs.glide.GlideApp
import com.travlog.android.apps.viewmodels.MyPageViewModel
import io.reactivex.android.schedulers.AndroidSchedulers
import kotlinx.android.synthetic.main.a_my_page.*
import javax.inject.Inject

class MyPageActivity : BaseActivity<MyPageViewModel>() {

    @Inject
    lateinit var viewModelFactory: ViewModelFactory

    override fun onCreate(savedInstanceState: Bundle?) {
        getAppInjector().inject(this)
        viewModel = getViewModel(viewModelFactory)

        super.onCreate(savedInstanceState)

        setContentView(R.layout.a_my_page)
        setSupportActionBar(toolbar)
        setDisplayHomeAsUpEnabled(true)

        username.clicks()
                .compose(bindToLifecycle())
                .subscribe { this.startSetUserNameActivity() }

        viewModel?.apply {
            outputs.setProfilePicture()
                    .compose(bindToLifecycle())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({ setProfilePicture(it) })

            outputs.setUsernameText()
                    .compose(bindToLifecycle())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({ setUsernameText(it) })
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean =
            menuInflater.inflate(R.menu.menu_settings, menu).run {
                return super.onCreateOptionsMenu(menu)
            }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_settings -> {
                startSettingsActivity()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun setProfilePicture(profilePicture: String) =
            GlideApp.with(this)
                    .load(profilePicture)
                    .apply(circleCropTransform())
                    .into(this.profile_picture)

    private fun setUsernameText(username: String) {
        this.username.text = username
    }

    private fun startSettingsActivity() =
            Intent(this, SettingsActivity::class.java).let {
                startActivityWithTransition(it, R.anim.slide_in_right, R.anim.fade_out_slide_out_left)
            }

    private fun startSetUserNameActivity() =
            Intent(this, SetUsernameActivity::class.java).let {
                startActivityWithTransition(it, R.anim.slide_in_right, R.anim.fade_out_slide_out_left)
            }
}
